File: java/src/main/java/cn/az/code/easy/ValidParentheses.java
package cn.az.code.easy;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * @author : Liz
 * @date : 2019/10/25
 **/
public class ValidParentheses {

    public static void main(String[] args) {
        System.out.println(valid("[[(]]{})"));
    }

    public static boolean valid(String strs) {
        Deque<Character> queue = new ArrayDeque<>();
        for (int i = 0; i < strs.length(); i++) {
            if (strs.charAt(i) == '(' || strs.charAt(i) == '[' || strs.charAt(i) == '{') {
                queue.add(strs.charAt(i));
            } else {
                if (queue.isEmpty()) {
                    return false;
                } else {
                    if (queue.peek() == '(' && strs.charAt(i) != ')') {
                        return false;
                    } else if (queue.peek() == '[' && strs.charAt(i) != ']') {
                        return false;
                    } else if (queue.peek() == '{' && strs.charAt(i) != '}') {
                        return false;
                    }
                    queue.pop();
                }
            }
        }
        return queue.isEmpty();
    }
}

File: archive/python/easy/ValidParentheses.py
def valid(s):
    lst = []
    for i in range(0, len(s)):
        if s[i] is '(' or s[i] is '[' or s[i] is '{':
            lst.append(s[i])
        else:
            if len(lst) is 0:
                return False
            else:
                if lst[len(lst) - 1] is '(' and s[i] is not ')':
                    return False
                elif lst[len(lst) - 1] is '[' and s[i] is not ']':
                    return False
                elif lst[len(lst) - 1] is '{' and s[i] is not '}':
                    return False
            lst.pop()
    return len(lst) is 0


if __name__ == '__main__':
    print(valid('[[]]'))
